### PR TITLE
Fix wrong OS regex in project converter

### LINKED ISSUE #73909: [4.0.rc5] Converter deletes lines that use OS.window_size
### Godot version

4.0.rc5.official

### System information

Windows 10

### Issue description

When converting a project from 3.5.1, the converter completely removes lines that use `OS.window_size` instead of upgrading them to a Godot 4 equivalent. Quite often this leads to errors, e.g. missing variable declarations, indented blocks with no opening `if` statement, etc.

### Steps to reproduce

1. Open the repro project in Godot 3.5.1 or earlier
2. Inspect contents of `demo_script.gd`
3. Close Godot 3.5.1 and run a full conversion to Godot 4
4. Open project in Godot 4
5. Observe that `demo_script.gd` is now in an error state due to lacking a variable declaration

### Minimal reproduction project

[OS Window Function Error.zip](https://github.com/godotengine/godot/files/10830465/OS.Window.Function.Error.zip)

### PR DESCRIPTION
**Pull request by KoBeWi**
Fixes #73909
The expression was using an optional capture, which is probably why it failed and removed the line if there was no `set_`/`get_`.

### CODE REVIEW

File: editor/project_converter_3_to_4.cpp [hunk 1]
```diff
@@ -78,11 +78,11 @@ class ProjectConverter3To4::RegExContainer {
 	RegEx reg_image_unlock = RegEx("([a-zA-Z0-9_\\.]+)\\.unlock\\(\\)");
 	RegEx reg_instantiate = RegEx("\\.instance\\(([^\\)]*)\\)");
 	// Simple OS properties with getters/setters.
-	RegEx reg_os_current_screen = RegEx("\\bOS\\.(set_|get_)?current_screen\\b");
-	RegEx reg_os_min_window_size = RegEx("\\bOS\\.(set_|get_)?min_window_size\\b");
-	RegEx reg_os_max_window_size = RegEx("\\bOS\\.(set_|get_)?max_window_size\\b");
-	RegEx reg_os_window_position = RegEx("\\bOS\\.(set_|get_)?window_position\\b");
-	RegEx reg_os_window_size = RegEx("\\bOS\\.(set_|get_)?window_size\\b");
+	RegEx reg_os_current_screen = RegEx("\\bOS\\.([set_|get_]?)current_screen\\b");
```
Comment (dalexeev) [inline editor/project_converter_3_to_4.cpp:81]:
```suggestion
	RegEx reg_os_current_screen = RegEx("\\bOS\\.((set_|get_)?)current_screen\\b");
```

**Comment by dalexeev** [comment]
This looks like a bug in `RegEx.sub()`. The original version should also work correctly, but for some reason it doesn't.

**Review by akien-mga — APPROVED** [review]

**Comment by dalexeev** [comment]
I'm worried this isn't the only place where our converter can silently delete lines. This can happen anywhere that has an optional capture groups and `$N`.

**Comment by akien-mga** [comment]
I'm a bit worried of doing a RegEx-wide change like #73973 so close to the 4.0 release, so I'd prefer going for ad hoc solutions like this one which only affects the converter.

The other optional capture groups I can find are:
```
	// CSharp keywords.
	RegEx keyword_csharp_remote = RegEx("\\[Remote(Attribute)?(\\(\\))?\\]");
	RegEx keyword_csharp_remotesync = RegEx("\\[(Remote)?Sync(Attribute)?(\\(\\))?\\]");
	RegEx keyword_csharp_puppet = RegEx("\\[(Puppet|Slave)(Attribute)?(\\(\\))?\\]");
	RegEx keyword_csharp_puppetsync = RegEx("\\[PuppetSync(Attribute)?(\\(\\))?\\]");
	RegEx keyword_csharp_master = RegEx("\\[Master(Attribute)?(\\(\\))?\\]");
	RegEx keyword_csharp_mastersync = RegEx("\\[MasterSync(Attribute)?(\\(\\))?\\]");
```
and
```
// Find "OS.property" OR "OS.get_property()" / "OS.is_property()".
static String make_regex_gds_os_property_get(String name, String get) {
	return String("\\bOS\\.(") + get + "_)?" + name + "(\\s*\\(\\s*\\))?";
}
```

**Comment by dalexeev** [comment]
What regular expression did you use to find these regular expressions? :smiley: Note that it can be not only `)?`, but also `)*`, and the capture group can be inside another subpattern.

I am not opposed to ad hoc solutions in this case. But maybe we should at least add error output to 4.0?

**Comment by akien-mga** [comment]
> What regular expression did you use to find these regular expressions? smiley Note that it can be not only `)?`, but also `)*`, and the capture group can be inside another subpattern.

`\)\?`, so yeah I probably missed a bunch :)

> But maybe we should at least add error output to 4.0? 

I really wouldn't want to take any risk at this stage. If there's any valid use case for the failure case, then it printing errors will be seen as a bug by users.

**Comment by akien-mga** [comment]
I'll merge this for now as it fixes a reported bug, and so far we haven't been told of other cases where the converter silently removes lines. But that's definitely something worth looking further into.

**Comment by dalexeev** [comment]
But if we return the original string in case of an error, and not an empty one, will the risks be minimal?

**Comment by akien-mga** [comment]
Maybe, but I would need to spend time figuring out what the exact possible error conditions and what would be the expected behavior in that case. Since currently users don't seem to complain about the lack of error nor the empty return, I'd err on the side of caution. It does make sense to me overall as a change for 4.1 but will need testing (and improving the unit tests to validate this).

**Pull request merged**